modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala

```diff
@@ -32,41 +34,36 @@ import weaver.scalacheck.Checkers
 
 object ConstructionServiceSuite extends MutableIOSuite with Checkers with TransactionGenerator {
 
-  type Res = (SecurityProvider[IO], KryoSerializer[F])
+  type Res = (SecurityProvider[IO], KryoSerializer[IO])
 
   override def sharedResource: Resource[IO, Res] =
     KryoSerializer
       .forAsync[IO](sharedKryoRegistrar)
       .flatMap(kryo => SecurityProvider.forAsync[IO].map(securityProvider => (securityProvider, kryo)))
 
-  def generateTestTransactions(wantSignedTransaction: Boolean)(implicit S: SecurityProvider[IO], K: KryoSerializer[IO]) = {
-    val getSignedTransaction = (transaction: Hashed[Transaction]) => transaction.signed
-    val getUnsignedTransaction = (transaction: Hashed[Transaction]) => transaction.signed.value
-
-    def txs(transactionConverter: Hashed[Transaction] => AnyRef) =
-      (KeyPairGenerator.makeKeyPair[IO], KeyPairGenerator.makeKeyPair[IO]).tupled.flatMap {
-        case (srcKey, dstKey) =>
-          val srcAddress = srcKey.getPublic.toAddress
-          val dstAddress = dstKey.getPublic.toAddress
-          val txCount = PosInt(100)
-
-          generateTransactions(srcAddress, srcKey, dstAddress, txCount)
-            .flatMap(_.traverse { hashedTransaction =>
-              KryoSerializer[F]
-                .serialize(transactionConverter(hashedTransaction))
-                .map(Hex.fromBytes(_))
-                .map((_, hashedTransaction.hash, hashedTransaction))
-                .liftTo[IO]
-            })
-      }
-
-    if (wantSignedTransaction) {
-      txs(getSignedTransaction)
-    } else {
-      txs(getUnsignedTransaction)
-    }
+  private def getBytes[G[_]: Sync](hashedTransaction: Hashed[schema.transaction.DAGTransaction], wantSignedTransaction: Boolean) = if (
+    wantSignedTransaction
+  ) {
+    Sync[G].delay(JsonSerializer.serialize(hashedTransaction.signed))
```

Comment:
`JsonSerializer.serialize` should return an effect type, otherwise it's a pure function. (this holds for the whole codebase, except interoperability with java or some external impure libraries). what's more - JsonSerializer has been added in this pull request, so you should know/check how it behaves. if it's not pure, then it should be fixed directly in the `JsonSerializer` not to publish impure interface outside.

Reply:
```
  def serialize[A: Encoder](content: A) =
    content.asJson.noSpaces.getBytes("UTF-8")
```
I had trouble bring the codebase down but I can see this definition, which has no return type. When I duplicate this in the IDE it returns `Array[Byte]`. When I follow the Circe code there is a place where they specifically say the one of the calls they use is unsafe.

So I agree, this (and the decode?) should return its result in an effect.

Reply:
and there should be a return type. I think I've already commented that in the original PR

Reply:
> btw. can you point me to the documentation/code which is unsafe? I cannot find it.

I'm having trouble retracing my steps :( I'll keep looking.
However, `getBytes` is a java function and it throws, so if we want to make `serialize` safer we need to something...

Reply:
:+1: I agree. @kriscieslak can you take a look?

Reply:
The delay has been removed.